### PR TITLE
fossa vps test

### PR DESCRIPTION
**Pull request by cnr**
This adds the `fossa vps test` command, which works similarly to `fossa test`, but checks for sherlock scan status instead of normal build status.

This also introduces a few small refactors:
1. `Data.Flag` module to replace using `Bool` for cli option parsing
2. a new `ApiOpts` datatype containing base url and api key for fossa-core endpoints
3. extracting `ApiOpts` out of the VPS god objects

### CODE REVIEW

File: src/App/Fossa/VPS/Report.hs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    Nothing -> do
+      sendIO $ threadDelay (pollDelaySeconds * 1_000_000)
+      waitForSherlockScan apiOpts locator scanId
+
+timeout
+  :: Int -- ^ number of seconds before timeout
+  -> IO a
+  -> IO (Maybe a)
+-- timeout seconds act = either id id <$> Async.race (Just <$> act) (threadDelay (seconds * 1_000_000) *> pure Nothing)
+timeout seconds act = either id id <$> Async.race (Just <$> act) (threadDelay (seconds * 1_000_000) $> Nothing)
```
Comment (scruffystuffs) [inline src/App/Fossa/VPS/Report.hs:121]: I think this is the third copy of this function.  Can we put it in `App.Fossa.API.BuildWait`?

File: src/App/Fossa/VPS/Test.hs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+pollDelaySeconds = 8
+
+waitForSherlockScan ::
+  (Has Diagnostics sig m, Has (Lift IO) sig m, Has Logger sig m) =>
+  ApiOpts ->
+  VPSCore.Locator ->
+  -- | scan ID
+  Text ->
+  m ()
+waitForSherlockScan apiOpts locator scanId = do
```
Comment (scruffystuffs) [inline src/App/Fossa/VPS/Test.hs:99]: Can we add this function to `BuildWait` and merge it with the on in `VPS.Report`?

File: src/App/Fossa/VPS/Report.hs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+pollDelaySeconds = 8
+
+waitForSherlockScan ::
+  (Has Diagnostics sig m, Has (Lift IO) sig m, Has Logger sig m) =>
+  ApiOpts ->
+  VPSCore.Locator ->
+  -- | scan ID
+  Text ->
+  m ()
+waitForSherlockScan apiOpts locator scanId = do
```
Comment (scruffystuffs) [inline src/App/Fossa/VPS/Report.hs:103]: Let's dedup this into `BuildWait`

File: src/App/Fossa/FossaAPIV1.hs [hunk 4]
```diff
@@ -325,7 +317,7 @@ instance FromJSON Issue where
     Issue <$> obj .: "id"
            <*> obj .:? "priorityString"
            <*> obj .: "resolved"
-           <*> obj .: "revisionId"
+           <*> obj .:? "revisionId" .!= "unknown project"
```
Comment (scruffystuffs) [inline src/App/Fossa/FossaAPIV1.hs:320]: [Question]: What's the background for this change?

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
Solid overall, 1 question and a few comments about deduplicating and sharing code.

File: src/Data/Flag.hs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+-- http://oleg.fi/gists/posts/2019-03-21-flag.html
```
Comment (scruffystuffs) [inline src/Data/Flag.hs:1]: This is such a weird trick.  But it's definitely a good move.

File: src/App/Fossa/FossaAPIV1.hs (hunk 4 above)
Comment (cnr) [reply src/App/Fossa/FossaAPIV1.hs:321]: Added a comment there; essentially, the shape of the response annoyingly changes when we ask for issues for a VPS project

**Review by scruffystuffs — APPROVED** [review]
Looks good, as always!

**Pull request merged**